sympy/core/exprtools.py

```diff
@@ -245,47 +246,50 @@ def decompose_power(expr):
     if exp.is_Number:
         if exp.is_Rational:
             if not exp.is_Integer:
-                base = Pow(base, Rational(1, exp.q))
-
-            exp = exp.p
+                base = Pow(base, Rational(1, cast(Rational, exp).q))
+            e = cast(Rational, exp).p
         else:
-            base, exp = expr, 1
+            base, e = expr, 1
     else:
         exp, tail = exp.as_coeff_Mul(rational=True)
 
         if exp is S.NegativeOne:
-            base, exp = Pow(base, tail), -1
+            base, e = Pow(base, tail), -1
         elif exp is not S.One:
-            tail = _keep_coeff(Rational(1, exp.q), tail)
-            base, exp = Pow(base, tail), exp.p
+            # todo: after dropping python 3.7 support, use overload and Literal
+            #  in as_coeff_Mul to make exp Rational, and remove these 2 casts
+            tail = _keep_coeff(Rational(1, cast(Rational, exp).q), tail)
+            base, e = Pow(base, tail), cast(Rational, exp).p
         else:
-            base, exp = expr, 1
+            base, e = expr, 1
 
-    return base, exp
+    return base, e
 
 
-def decompose_power_rat(expr):
+def decompose_power_rat(expr: Expr) -> tTuple[Expr, Rational]:
     """
     Decompose power into symbolic base and rational exponent.
 
     """
     base, exp = expr.as_base_exp()
```

Comment:
maybe change this exp to e, too?

Reply:
Not applicable. `exp` is `Expr`, but e the return value is `Rational`. mypy complains:
error: Incompatible return value type (got "Tuple[Expr, Expr]", expected "Tuple[Expr, Rational]")